// noa-engine/src/components/entmesh.js

var vec3 = require('gl-vec3')


export default function (noa) {
    return {

        name: 'entmesh',

        order: 100,

        state: {
            mesh: null,
            offset: null,
			moving:false,
			fired:false,
			walkcycle:0,
			health: 1000,
			dead:false,
			far:false,
			npcList:null,
			playerMesh:null,
			mob:false,
			thisSocket:null,
			leash:false,
			leashmesh:null
			
        },


        onAdd: function (eid, state) {
			state.playerMesh=noa.ents.getState(noa.playerEntity, noa.entities.names.mesh).mesh
            // implicitly assume there's already a position component
            var posDat = noa.ents.getPositionData(eid)//
            if (state.mesh) {
				
				
                noa.rendering.addMeshToScene(state.mesh, false, posDat.position)
				if(state.mesh.getChildren()>0){
					for (var i=0;i<state.mesh._children[0].getChildren().length;i++){
						 noa.rendering.addMeshToScene(state.mesh._children[0]._children[i], false, posDat.position)//
					}
				}
					
				for (var i=0;i<state.mesh.getChildren().length;i++){
					 noa.rendering.addMeshToScene(state.mesh._children[i], false, posDat.position)
				}
	
			
				
            } else {
                throw new Error('Mesh component added without a mesh - probably a bug!')
            }
            if (!state.offset) state.offset = new vec3.create()

            // set mesh to correct position
            var rpos = posDat._renderPosition
            state.mesh.position.copyFromFloats(
                rpos[0] + state.offset[0],
                rpos[1] + state.offset[1],
                rpos[2] + state.offset[2])
        },


        onRemove: function (eid, state) {
            state.mesh.dispose()
        },



        renderSystem: function (dt, states) {
            // before render move each mesh to its render position, 
            // set by the physics engine or driving logic

            states.forEach(state => {
                var id = state.__id
				
				
				
				
		/*	if(state.leash!==null){
				
				var pos=noa.ents.getState(id, 'position').position
	
	
	var c=	getdistance(pos,state.leash)
 
var customMesh  = BABYLON.MeshBuilder.CreatePlane('sprite-',{height:c,width:0.1}, scene)
	
var mat = new BABYLON.StandardMaterial("mat", scene);
mat.diffuseColor = BABYLON.Color3.White();
mat.ambientColor = BABYLON.Color3.White();
	mat.backFaceCulling = false;
	//mat.diffuseTexture=(mod+'textures/block/sand.png',scene)
	customMesh.material = mat;
					
				//console.log(c);
					
					function getdistance (playerMesh,playerMesh2){
                                        var dblDistanceX = playerMesh[0] - playerMesh2[0];
										var dblDistanceY =  playerMesh[1] - playerMesh2[1];
										var dblDistanceZ =  playerMesh[2] - playerMesh2[2];
										
								var realDistance=Math.sqrt((dblDistanceX * dblDistanceX) + (dblDistanceY * dblDistanceY) + (dblDistanceZ * dblDistanceZ));
								return  realDistance;
							}
							
							
							  function lookat(objectlooker,objecttarget){
	
						var direction=Math.atan2(objectlooker[0]-objecttarget[0],objectlooker[2]-objecttarget[2])
						return direction;
					}
				
				noa.rendering.addMeshToScene(customMesh,false)
				customMesh.rotation.y=lookat(pos,state.leash)
				customMesh.rotation.x=lookat([pos[0],pos[1]+1,pos[2]],state.leash)
				customMesh.position.x=state.leash[0]-0.5
				customMesh.position.y=state.leash[1]
				customMesh.position.z=state.leash[2]-0.5
				state.leash=null
				
				setInterval(function(){ 
				customMesh.scaling.z=getdistance(pos,state.leash)
                  //state.leash=noa.ents.getState(noa.playerEntity, 'position').position
				 // customMesh.dispose()
				}, 1000);
				
				
			}*/
				
					if(state.dead){
						
						for (const key in state.npcList) {
    
								if(state.npcList[key]==id){
								
									//delete state.npcList[key]
									//state.dead=true
									
									state.thisSocket.emit('despawn', key)
									
									//noa.ents.getState(data.id, 'mesh2').dead=true
								}
					  }
					 noa.entities.deleteEntity(id)
					}
					
					
					
				var c=	getdistance(state.playerMesh,state.mesh)
				//console.log(c);
					
					function getdistance (playerMesh,playerMesh2){
                                        var dblDistanceX = playerMesh.position.x - playerMesh2.position.x;
										var dblDistanceY = playerMesh.position.y - playerMesh2.position.y;
										var dblDistanceZ = playerMesh.position.z - playerMesh2.position.z;
										
								var realDistance=Math.sqrt((dblDistanceX * dblDistanceX) + (dblDistanceY * dblDistanceY) + (dblDistanceZ * dblDistanceZ));
								return  realDistance;
							}
					
					if(c>200){//was 40
						
						for (const key in state.npcList) {
    
								if(state.npcList[key]==id){
									
									
								state.thisSocket.emit('despawn', key)
									//delete state.npcList[key]
									//state.dead=true
									
									//noa.ents.getState(data.id, 'mesh2').dead=true
								}
					  }
						
					}
		
	
				
				
				

                var rpos = noa.ents.getPositionData(id)._renderPosition
                state.mesh.position.copyFromFloats(
                    rpos[0] + state.offset[0],
                    rpos[1] + state.offset[1],
                    rpos[2] + state.offset[2])
            })
        }
		
		


    }
}
