// noa-engine/src/components/projmesh.js

var vec3 = require('gl-vec3')


export default function (noa) {
    return {

        name: 'projmesh',

        order: 100,

        state: {
            mesh: null,
            offset: [0,0.5,0],
			walkcycle:0,
			fired:false,
			walkcyle:0,
			headtimer:0,
			thirdp:false,
			moving:false
			
			
        },


        onAdd: function (eid, state) {
			
            // implicitly assume there's already a position component
            var posDat = noa.ents.getPositionData(eid)
            if (state.mesh) {
                noa.rendering.addMeshToScene(state.mesh, false, posDat.position)
				
				 noa.rendering.addMeshToScene(state.mesh._children[0], false, posDat.position)
				 noa.rendering.addMeshToScene(state.mesh._children[1], false, posDat.position)
				
					// noa.rendering.addMeshToScene(state.mesh._children[0], false, posDat.position)
				
				
				//var body = noa.ents.getPhysicsBody(eid)
            } else {
                throw new Error('Mesh component added without a mesh - probably a bug!')
            }
            if (!state.offset) state.offset = new vec3.create()

            // set mesh to correct position
            var rpos = posDat._renderPosition
            state.mesh.position.copyFromFloats(
                rpos[0] + state.offset[0],
                rpos[1] + state.offset[1],
                rpos[2] + state.offset[2])
        },


        onRemove: function (eid, state) {
            state.mesh.dispose()
        },



        renderSystem: function (dt, states) {
            // before render move each mesh to its render position, 
            // set by the physics engine or driving logic

            states.forEach(state => {
                var id = state.__id
				
			

                var rpos = noa.ents.getPositionData(id)._renderPosition
                state.mesh.position.copyFromFloats(
                    rpos[0] + state.offset[0],
                    rpos[1] + state.offset[1],
                    rpos[2] + state.offset[2])
            })
        }
		
		


    }
}
